Wait until action

xiyuoh opened the pull request:
This PR provides a `wait_until` PerformAction plugin for the MiR fleet adapter. It depends on and follows https://github.com/open-rmf/fleet_adapter_mir/pull/36.

This performable action enables robots to travel to a waypoint and wait there until a timeout or a move off signal is received. The expected task behavior would be:
1. Robot goes to a waypoint and waits for X seconds (configurable)
2. If the robot receives a move-off signal (configurable), it will end the waiting action and move to the next waypoint
3. Robot repeats Steps 1. and 2. with subsequent waypoints on a list until it has travelled to all the waypoints in the task.

During the waiting process, the robot may only move off if the default timeout has been reached, or if it receives a move off signal. This move off signal can be configured by the user depending on the user case. The plugin provides 2 out-of-the-box move off signals as well as the option for users to customize their own.
- `mission`: The waiting action would be mission-based, i.e. the fleet adapter would trigger a MiR mission specified in the config when the robot starts waiting, and end the waiting action when the mission is completed.
- `plc`: The waiting action would begin and the fleet adapter monitors the state of the specified PLC register. Once it turns to a positive integer or `True`, the waiting action ends.
- `custom`: Users can define their own move off signal by writing their own `MoveOff` module and provide it to the fleet config. A abstract class `BaseMoveOff` and example of an implementation are provided in `rmf_move_off.py` and `rmf_move_off_on_alert.py` respectively.

For cases where users might want the move off signal to vary from task to task within the same fleet, they may provide the move off behavior in their task description instead of the fleet config. Any move off signal config provided in a task description will override the plugin config. If neither the task description or the plugin config provided sufficient information, no move off signals will be configured and the robot will simply wait for the full duration of the timeout.

The README contains greater detail of how customization for this action should work.

TODO:
- [ ] Test in sim

xiyuoh committed:
EFC fleet adapter

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Create standard RMF missions from fleet adapter

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Set up plugin system for custom MiR perform actions

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Cart delivery plugin fixes

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add cart pickup task script

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Footprint and marker type things and update readme

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Restore mir/mirfm comparison and add note that mirfm is not actively supported

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Make marker type a variable when posting missions

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Update dispatch pickup cancellation behavior and configs

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Remove limits on charging mission

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Fixes bugs for cancelled pickup states, refactor mission info stored in pickup/dropoff, and move action-related code into its own package

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Use ActionFactory to create MirAction objects

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Remove fleet_adapter_mir_actions as dependency in fleet_adapter_mir

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Move MirAction to fleet_adapter_mir pkg

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Working plugin module via importlib and modified dispatch_delivery task

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Rename actions arg parsed to fleet_adapter_mir to rmf_missions

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add module key to config

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Remove footprint update from cart delivery plugin and cancel task if
pickup pos is not found

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Configurable transformation params

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add in retry mechanism for navigate cmds

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add retry mechanism for dock and go to known position

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Fix wrong input argument for task cancellation

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Separate cart detection related functions into another plugin

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add example for module config and minor cleanup

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Codestyle cleanup

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Update README with instruction to write own CartDetection module

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Remove position_put logic when localizing to prevent overwriting existing MiR coords

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add disconnect mutex

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Check config for lift or localize positions before requesting for localize missions

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Style

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Cleanup and add license declaration

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add WaitUntil perform action with mission and PLC move off behavior

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Cleanup

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Merge branch 'xiyu/efc_fleet_adapter_mir' into xiyu/wait_until_action

xiyuoh committed:
Custom move off behavior with examples

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add wait_until task

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add readme and rename task to multistop

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Remove duplicate instructions

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Merge branch 'main' into xiyu/wait_until_action

xiyuoh committed:
Update wait until action

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Merge branch 'main' into xiyu/wait_until_action

xiyuoh committed:
Refactor signal type configuration

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Codestyle - break lines that are too long

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add examples

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Add remaining examples

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Collapse example diagrams

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

luca-della-vedova commented on configs/mir_config.yaml:
```diff
@@ -111,3 +111,19 @@ plugins:
+  rmf_wait_until:
+    module: "fleet_adapter_mir_actions.rmf_wait_until"
+    actions: ["wait_until"]
+    timeout: 1800 # Optional, in seconds. Timeout if there is no move-off signal received, after which robot will move off regardless. Default to 60 seconds.
```
Nit, if we are providing a default config can we make sure the default value is the same as the codebase? So either change this to `timeout: 60` or the codebase to have a default value of `1800`? There are two different defaults and I find that a bit confusing. All the other parameter values in the file seem to be consistent with what the default is (`retry_count`, `update_gap`, `resubmit_on_abort`)

luca-della-vedova commented on docs/mir_actions.md:
```diff
@@ -54,3 +55,111 @@ ros2 run fleet_adapter_mir_tasks dispatch_delivery -g go_to_waypoint -p pickup_l
+
+### Overview
+
+The `rmf_wait_until` plugin introduces the `wait_until` action. It allows users to command robots to wait at a specified location until it receives a move off signal or until a configured timeout. The robot would then be free to move on to carry out the remainder of its task, or complete the task and proceed to its idle behavior. During this waiting period, the user may command the robot to perfo ...
```
Nit `to command robots [...] until it`, change to either all singular or all plural

luca-della-vedova commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,473 @@
+                raise ValueError(
+                    f'User provided a default signal type in the action '
+                    f'config for WaitUntil MirAction, but the default signal '
+                    f'type is not configured!')
```
Nit, log in the exception what the user actually provided to make it more helpful

luca-della-vedova commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_move_off.py:
```diff
@@ -0,0 +1,65 @@
+        # To be implemented
+        ...
+
+    # --------------------------------------------------------------------------
+    # HELPFUL FUNCTIONS FOR INTERACTING WITH MIR REST API
+    # --------------------------------------------------------------------------
+
+    def register_get(self, register: int):
+        if not self.context.api.connected:
+            return None
+        try:
+            response = requests.get(
+                self.context.api.prefix + f'registers/{register}',
+                headers=self.context.api.headers,
+                timeout=self.context.api.timeout)
+            if self.context.api.debug:
+                print(f"Response: {response.headers}")
+            # Response value is string, return integer of value
+            return int(response.json().get('value', 0))
+        except HTTPError as http_err:
+            print(f"HTTP error: {http_err}")
+            return None
+        except Exception as err:
+            print(f"Other  error: {err}")
+            return None
```
Is this used? I tried to look for `register_get` and I could only find a duplicated definition and a single call of it in the `rmf_wait_until` module

xiyuoh replied:
This is provided in case users need to check the PLC register as part of their custom move off module. But since it is provided in the `rmf_wait_until` module it probably makes more sense to remove it from the abstract class.

xiyuoh replied:
[e9c0700](https://github.com/open-rmf/fleet_adapter_mir/pull/39/commits/e9c0700b918ac553cdbdd6a5dbe2a4c85ff3d499)

luca-della-vedova commented on fleet_adapter_mir_tasks/fleet_adapter_mir_tasks/dispatch_multistop.py:
```diff
@@ -0,0 +1,216 @@
+
+        # GoToPlace + wait
+        for i in range(len(self.args.go_to)):
+            place = self.args.go_to[i]
```
Nit, can just refactor to `for place in self.args.go_to:`

luca-della-vedova commented on fleet_adapter_mir_tasks/fleet_adapter_mir_tasks/dispatch_multistop.py:
```diff
@@ -0,0 +1,216 @@
+                "description": {
+                    "unix_millis_action_duration_estimate": 60000,
+                    "category": 'wait_until',
+                    "description": {
+                        "timeout": self.args.timeout,  # seconds
+                        "update_gap": self.args.update_gap,  # seconds,
+                        "signal_type": signal_type,
+                        "signal_config": signal_config
+                    }
+                }
+            }]
+            if self.args.timeout is not None:
+                wait_activity[0]['description']['description']['timeout'] = \
+                    self.args.timeout
+            if self.args.update_gap is not None:
+                wait_activity[0]['description']['description']['update_gap'] = \
+                    self.args.update_gap
```
Is the assignment for `timeout` and `update_gap` here duplicated?  I see both a:

```
"timeout": self.args.timeout
[...]
if self.args.timeout is not None:
  wait_activity[0]['description']['description']['timeout'] = self.args.timeout
```

Did you mean to specify a default in the first assignment when you create the `wait_activity`?

xiyuoh replied:
Ah I forgot to remove the fields when initializing description. The intent is to only add them in if the param was provided, so if the WaitUntil action doesn't find the `timeout`/`update_gap` from the task description, they can use the default configured values.

xiyuoh replied:
[e9c0700](https://github.com/open-rmf/fleet_adapter_mir/pull/39/commits/e9c0700b918ac553cdbdd6a5dbe2a4c85ff3d499)

luca-della-vedova commented on configs/mir_config.yaml:
```diff
@@ -111,3 +111,19 @@ plugins:
+    update_gap: 30 # Optional, in seconds. The fleet adapter will log an update using this interval. Default to 30 seconds.
+    signal_type: # Defines the default configs for various move off signal types. If this field is not provided, the robot will wait the full duration of the default timeout. Signal configs can be overridden by task description.
+      # Currently three signal types are supported: [mission, plc, custom].
+      mission:
+        mission_name: "some_mission_name" # If this mission is found on the robot, it will be queued when the action starts, and the robot will move off when the mission is completed.
+        retry_count: 10 # Optional field for signal type "mission". The fleet adapter will re-attempt queueing the mission for this number of tries.
+        resubmit_on_abort: False # Optional field for signal type "mission". Set to True to resubmit mission if the mission gets aborted by the robot. Default to False.
+      plc:
+        register: 20 # Fill in with PLC register number. Robot moves off when this PLC register returns True
+      custom:
+        module: "fleet_adapter_mir_actions.rmf_move_off_on_alert" # Fill in with path to custom module written. Robot moves off when module provides the user-defined move off signal.
+      default: "custom" # Specifies a default signal type out of those listed above. If the task description does not provide any signal config, the fleet adapter will use this default signal type.
```
This is more something for discussion rather than a request for change. Thinking again about the limitation that the adapter will only support one `mission`, one `plc` and one `custom`, which is especially important for `custom` since the `module` is not exposed to the user API, what do you think about a config file like the following:

```
signals:
  sample_mission: # Only a unique name
    signal_type: mission # Specifies the actual type
    mission_name: "some_mission_name"
    [,,,]
  sample_mission_2:
    signal_type: mission # Specifies the actual type
    mission_name: "some_other_mission_name"
    [,,,]
  custom_1:
     signal_type: custom
     module: "fleet_adapter_mir_actions.rmf_move_off_on_alert"
  custom_2:
     signal_type: custom
     module: "fleet_adapter_mir_actions.rmf_move_off_on_something_else"
default_signal: "custom_1"
```

In theory this should allow multiple entries which can be addressed by name (since the list is a map), but to be honest I don't know how common the use case of having such level of configurability is and if it justifies the added complexity, what do you think?

xiyuoh replied:
This is a really interesting suggestion and can be incorporated very easily. I can see the main pro being having everything configured at launch, such that users who would like 10 different missions to be triggered for various tasks won't have to constantly overwrite default values with their task descriptions. In fact, this helps to reduce the complexity of task descriptions - instead of allowing users to pick and choose the values they want to overwrite, I'd limit their options to either A) pick an existing configured signal type, B) provide a mission name/PLC register (no more submitting "PLC" task without specifying the register value), or C) pick the default signal like before.

From experience I feel like high level of configurability never hurts (as long as they are clear), and more often than not it is requested when we least expect it 🤣  I'll work on adding this.

xiyuoh replied:
[6284a28](https://github.com/open-rmf/fleet_adapter_mir/pull/39/commits/6284a28aaad8aacdf6c916a3861128606fa315f9)

luca-della-vedova reviewed (commented):
With a note that I didn't run this since it requires hardware and I only did the test of running it in my brain, I just have a few small comments here and there and a more meaty point where I'd be happy to hear your input. Generally speaking the quality of the coding and the API looks quite solid to me

xiyuoh committed:
Address comments

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Enable multiple signal config of each type

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Update docs

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

aaronchongth commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,483 @@
+                f'will have to be provided in the task description submitted '
+                f'to RMF.'
+            )
+            return
```
nit: this can be returned early for `if signals is None:`, so that we have less nested if else statements

aaronchongth commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,483 @@
+                f'populated in the task description, there will be no move '
+                f'off signal available and the robot will wait for the full '
+                f'duration of the timeout when this action is triggered.'
+            )
```
nit: same here, can be returned early for less nested statements

aaronchongth commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,483 @@
+            case 'wait_until':
+                return True
+            case _:
+                return False
```
nit: not to overuse match-case statements where unnecessary
```suggestion
        return True if category == 'wait_until' else False
```

aaronchongth commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,483 @@
+            case 'wait_until':
+                return WaitUntil(
+                    description, execution, self.context,
+                    self.custom_modules)
```
Can be handled with an if case

hmmm, on the other hand, since we expect this function to only return a `MirAction`, if we receive an invalid `category`, should we raise an exception? I would probably prefer to return a `None`, and let upstream handle it more elegantly, but I see that the base class signature has been decided on. With all the initial checks in place, it's highly unlikely that we'd get an invalid category, so if it does happen, it might indicate something very wrong has happened and an exception might be a good thing to raise

aaronchongth commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,483 @@
+                headers=self.context.api.headers,
+                timeout=self.context.api.timeout)
+            if self.context.api.debug:
+                print(f"Response: {response.headers}")
```
Should we use `self.context.node.get_logger().debug(` instead? or `info(`

applies to the other print statements below too

aaronchongth commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,483 @@
+                    f'with mission queue id {mission_queue_id}, marking '
+                    f'action as completed.'
+                )
+                return True
```
nit: can be returned earlier to prevent nesting

aaronchongth commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,483 @@
+            self.context.node.get_logger().info(
+                f'Robot [{self.context.name}] has completed waiting for '
+                f'{self.wait_timeout} seconds without move off signal, '
+                f'action as complete.'
```
```suggestion
                f'marking action as complete.'
```

aaronchongth reviewed (changes requested):
This looks great! I've only briefly gone through the rest of the code base, so can't comment much about architecture and design decisions, but I'm definitely digging the plugin approach. I've left a few comments that are generally just nitpicks, let me know what you think

xiyuoh committed:
Review comments

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Reduce line

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh commented:
@aaronchongth Thanks for the review! Addressed them in [218db4d](https://github.com/open-rmf/fleet_adapter_mir/pull/39/commits/218db4da5e30acc0bab707ffefc5919131b94ecb). Extended some of them to `rmf_cart_delivery` and `rmf_cart_detection` as well.

aaronchongth reviewed (approved):
Looks good from my end!

mxgrey commented on configs/mir_config.yaml:
```diff
@@ -111,3 +111,31 @@ plugins:
+  rmf_wait_until:
+    module: "fleet_adapter_mir_actions.rmf_wait_until"
+    actions: ["wait_until"]
+    timeout: 60 # Optional, in seconds. Timeout if there is no move-off signal received, after which robot will move off regardless. Default to 60 seconds.
```
A bit of a nitpick, I believe this is a default or fallback timeout in the event the `wait_until` action didn't specify a timeout in its description. I think we should reflect that in the name here, e.g. call it `default_timeout`.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,485 @@
+            )
+            return
+
+        if default_signal not in supported_signal_types:
```
This check doesn't make sense to me. I expect `default_signal` to be the name of a key in the `signals` dictionary, not the name of a signal type (i.e. `mission`, `plc`, or `custom`) which is what `supported_signal_types` would contain.

It's especially strange that the `elif` then tests the string against `signals.keys()`.. that implies that the string of the default signal must match both (1) one of the supported signal type names, and (2) one of user-define signal names. Anything else would raise an exception.

I'm guessing the logic was supposed to be something like this:

```python
if default_signal not in signals.keys():
    ...
elif signals[default_signal]['signal_type'] not in supported_signal_types:
    ...
```

But even in this case, the `elif` should be redundant with the earlier verification of all signals.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,485 @@
+                f'default signal type is not configured!')
+
+    def supports_action(self, category: str) -> bool:
+        return True if category == 'wait_until' else False
```
```suggestion
        return category == 'wait_until'
```

mxgrey commented on configs/mir_config.yaml:
```diff
@@ -111,3 +111,31 @@ plugins:
+        signal_type: "plc" # Specifies the actual type
+        register: 21
+      custom_1:
+        signal_type: "custom" # Specifies the actual type
```
You could consider calling this `signal_type: "plugin"` instead of `custom` to make it more explicit that it uses a plugin system. But I have no objection to keeping it as `custom`.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,485 @@
+
+            # Pass signal config into verify functions to validate types
+            # and catch missing values
+            match signal_type:
```
I notice there are many places in this code where it does `match signal_type`.

Since we already support custom signal types via plugins, I would've expected that we could use the base class (I think `MoveOff`) to define all the interfaces needed across different signal types. With that we could have

```python
class WaitForMission(MoveOff):
    ...

class WaitForPLC(MoveOff):
    ...
```

and we would only need this one `match signal_type:` to instantiate `self.signals[signal_name] = ` for whichever signal type was asked for.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,485 @@
+                )
+                signal_cb = lambda: self.check_plc_register(register)
+            case 'custom':
+                module = custom_modules.get(signal_type)
```
Looks like this line is always equivalent to 

```python
module = custom_modules.get('custom')
```

which doesn't seem right to me. I think this wants to be

```python
module = custom_modules.get(signal_config['module'])
```

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_wait_until.py:
```diff
@@ -0,0 +1,485 @@
+        # in the task description overrides the default config provided in
+        # action config.
+        if 'signal_type' in description:
+            signal_type = description['signal_type']
```
The `signal_type` terminology is confusing here, because elsewhere in this package "signal type" refers to one of `mission`, `plc`, or `custom`, but here I believe it's meant to be one of the key names inside the `signals` dictionary. I think the resulting confusion may actually be the cause of some bugs later in this function.

I would suggest we call this `signal_name` instead of `signal_type`, including in the `description`.

xiyuoh replied:
I think you're spot on, the terminology introduced was definitely unclear. While addressing an above comment I moved the PLC and mission logic into their own plugins and removed the need for `signal_type` altogether, I think the variable names are less confusing now.

mxgrey reviewed (changes requested):
I've left a few suggestions that I think will reduce confusion and improve maintainability. I also believe I've spotted a few bugs that we should definitely address before merging.

xiyuoh committed:
Fix signal bugs

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Use different term for fleet configured signals

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Move PLC and mission to MoveOff plugins

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Remove unused diagrams

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Minor config fix

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh commented:
Thanks @mxgrey for pointing out the bugs! I've fixed them and also implemented plugins for PLC and mission signals, removing the need for `custom` signal types (and signal types in general). Things should be less cluttered now, ready for another round of review.

mxgrey reviewed (approved):
Sorry for taking so long to get back to this. This is looking conceptually very good now with the last refactoring that you did.

I'll go ahead and approve, but given that it's Python, I'm slightly concerned there could be subtle runtime exceptions lurking in the shadows. I wonder if we can find anyone in the community willing to test this directly with MiR hardware before merging. I'll reach out at the upcoming community events.

xiyuoh committed:
Merge branch 'main' into xiyu/wait_until_action

xiyuoh committed:
Make things work

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>

xiyuoh committed:
Minor misc fixes

Signed-off-by: Xiyu Oh <xiyu@openrobotics.org>
